src/main/java/net/minecraftforge/oredict/ShapedOreRecipe.java

```diff
@@ -143,6 +147,41 @@ protected boolean checkMatch(InventoryCrafting inv, int startX, int startY, bool
         return true;
     }
 
+    @Override
+    public NonNullList<ItemStack> getRemainingItems(InventoryCrafting inv)
+    {
+        NonNullList<ItemStack> remainders = NonNullList.withSize(inv.getSizeInventory(), ItemStack.EMPTY);
+        for (int index = 0; index < remainders.size(); index++)
+        {
+            ItemStack itemStack = inv.getStackInSlot(index);
+            if (!itemStack.isEmpty())
+            {
+                Ingredient ingredient = input.get(index);
+                if (ingredient instanceof FluidIngredient)
```

Comment:
Could we make this an interface that fluid ingredients implement and use a callback for this logic? I could see this being useful for an RF battery ingredient to have a callback to remove energy, or other similar cases where the ingredient can define special logic on crafting with it

Reply:
Could that be a part of `Ingredient` rather than an interface?

Reply:
Well, we would need something to check for if we use the default logic of `getContainerItem()`, though I guess that logic could just be placed in ingredient itself.

Reply:
After some further thinking, I think it's better to be a separate pull request - for sake of making things clear.

Reply:
@KnightMiner I'd suggest adding a method in ingredient class to handle remaining items instead.

Reply:
I'd personally would just rename this PR and include it as I think the better customizability would be nice, but if you would rather do it separately that's fine (or I might try doing it later)

Reply:
@KnightMiner I think I changed my mind. `Ingredient` should be a black-box IMHO. Will kickstart on such an interface...

Reply:
Problem with doing that is that it requires re-working how the remaining items are done and re-doing the matching system to align the ingredient to the input.
This is part of the reason i've written my matcher code to return the stuff that it does. But From my discussions with various people my idea is dumb and not the way they want it to go.. so I haven't implemented it.